WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadService.java

```diff
@@ -0,0 +1,448 @@
+package org.wordpress.android.ui.uploads;
+
+import android.app.Service;
+import android.content.Context;
+import android.content.Intent;
+import android.os.IBinder;
+import android.support.annotation.NonNull;
+
+import org.greenrobot.eventbus.Subscribe;
+import org.greenrobot.eventbus.ThreadMode;
+import org.wordpress.android.R;
+import org.wordpress.android.WordPress;
+import org.wordpress.android.fluxc.Dispatcher;
+import org.wordpress.android.fluxc.generated.PostActionBuilder;
+import org.wordpress.android.fluxc.model.MediaModel;
+import org.wordpress.android.fluxc.model.PostModel;
+import org.wordpress.android.fluxc.model.SiteModel;
+import org.wordpress.android.fluxc.store.MediaStore;
+import org.wordpress.android.fluxc.store.MediaStore.OnMediaUploaded;
+import org.wordpress.android.fluxc.store.PostStore;
+import org.wordpress.android.fluxc.store.PostStore.OnPostUploaded;
+import org.wordpress.android.fluxc.store.SiteStore;
+import org.wordpress.android.ui.media.services.MediaUploadReadyListener;
+import org.wordpress.android.util.AppLog;
+import org.wordpress.android.util.AppLog.T;
+import org.wordpress.android.util.DateTimeUtils;
+import org.wordpress.android.util.FluxCUtils;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.inject.Inject;
+
+import de.greenrobot.event.EventBus;
+
+public class UploadService extends Service {
+    private static final String KEY_MEDIA_LIST = "mediaList";
+    private static final String KEY_LOCAL_POST_ID = "localPostId";
+    private static final String KEY_SHOULD_TRACK_ANALYTICS = "shouldTrackPostAnalytics";
+
+    private MediaUploadHandler mMediaUploadHandler;
+    private PostUploadHandler mPostUploadHandler;
+    private PostUploadNotifier mPostUploadNotifier;
+
+    // Posts that we're withholding from the PostUploadManager until their pending media uploads are completed
+    private static final List<PostModel> sPostsWithPendingMedia = new ArrayList<>();
+
+    // To avoid conflicts by editing the post each time a single upload completes, this map tracks completed media by
+    // the post they're attached to, allowing us to update the post with the media URLs in a single batch at the end
+    private static final ConcurrentHashMap<Integer, List<MediaModel>> sCompletedMediaByPost = new ConcurrentHashMap<>();
+
+    @Inject Dispatcher mDispatcher;
+    @Inject MediaStore mMediaStore;
+    @Inject PostStore mPostStore;
+    @Inject SiteStore mSiteStore;
+
+    @Override
+    public void onCreate() {
+        super.onCreate();
+        ((WordPress) getApplication()).component().inject(this);
+        AppLog.i(T.MAIN, "UploadService > Created");
+        mDispatcher.register(this);
+        // TODO: Recover any posts/media uploads that were interrupted by the service being stopped
+    }
+
+    @Override
+    public void onDestroy() {
+        mMediaUploadHandler.cancelInProgressUploads();
```

Comment:
we should check `mMediaUpoadHandler != null`  before accessing its methods here, as they are only created in `onStartCommand`, I believe there's this slight possibility `onCreate` is called and then `onDestroy` gets called immediately afterwards (without `onStartCommand` being called).